File: src/main/java/de/pflugradts/pwman3/domain/model/password/Password.java
package de.pflugradts.pwman3.domain.model.password;

import de.pflugradts.pwman3.domain.model.transfer.Bytes;
import de.pflugradts.pwman3.domain.model.ddd.DomainEntity;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;

@EqualsAndHashCode
final class Password implements DomainEntity {

    @Getter(AccessLevel.PRIVATE)
    private Bytes bytes;

    private Password(final Bytes bytes) {
        this.bytes = bytes.copy();
    }

    static Password create(final Bytes bytes) {
        return new Password(bytes);
    }

    void update(final Bytes newBytes) {
        this.bytes = newBytes.copy();
    }

    void discard() {
        this.getBytes().scramble();
    }

    Bytes view() {
        return getBytes().copy();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/model/egg/Password.kt
package de.pflugradts.passbird.domain.model.egg

import de.pflugradts.passbird.domain.model.ddd.DomainEntity
import de.pflugradts.passbird.domain.model.shell.EncryptedShell

class Password private constructor(private var encryptedShell: EncryptedShell) : DomainEntity {
    fun update(newEncryptedShell: EncryptedShell) {
        encryptedShell = newEncryptedShell.copy()
    }
    fun discard() = encryptedShell.scramble()
    fun view() = encryptedShell.copy()
    override fun equals(other: Any?) = (other as? Password)?.view() == encryptedShell
    override fun hashCode() = encryptedShell.hashCode()
    companion object {
        fun createPassword(encryptedShell: EncryptedShell) = Password(encryptedShell.copy())
    }
}
